pinecone/core/grpc/index_grpc.py

```diff
@@ -317,12 +332,56 @@ def upsert(self,
                              'To upsert in parallel, please follow: '
                              'https://docs.pinecone.io/docs/performance-tuning')
 
+        def _dict_to_grpc_vector(item):
+            item_keys = set(item.keys())
+            if not item_keys.issuperset(REQUIRED_VECTOR_FIELDS):
+                raise ValueError(
+                    f"Vector dictionary is missing required fields: {list(REQUIRED_VECTOR_FIELDS - item_keys)}")
+
+            excessive_keys = item_keys - (REQUIRED_VECTOR_FIELDS | OPTIONAL_VECTOR_FIELDS)
+            if len(excessive_keys) > 0:
+                warnings.warn(f"Found excessive keys in the vector dictionary: {list(excessive_keys)}. "
```

Comment:
nit: "Found excess keys"

Also, is this a new usage of the python warnings api? IIRC users need to to programmatically enable "warning output" to see these; we should make sure we have a unified strategy between this and adding warnings as debug logs to the "pinecone.*" logger - and we should probably add some docs to the [Troubleshooting](https://docs.pinecone.io/docs/troubleshooting) page about this

Reply:
Actually, this inputting vectors as dicts is a net new api in the client, right? It's probably preferable to make this a ValueError and not proceed with the request at all. I'd rather we fail fast even in a few rare cases where a user includes a key which is truly irrelevant rather than silently accepting e.g. upserts with a "metadat" key instead of "metadata" and leaving the user to debug and figure out why their metadata isn't coming back to them on reads.

Reply:
(This same UX issue is actively biting us in the json api)

Reply:
Agree, keeps it consistent with the tuple variant where we are throwing a value error on more than 3 items

Reply:
Totally agree, will change to `raise`.  

BTW `warnings` should be printed (to `stderr`) by default like any other exception (the only difference is they don't stop execution)